bridge/engine/EngineMixer.cpp

```diff
@@ -4014,143 +4061,148 @@ void EngineMixer::onBarbellUserMediaMap(size_t barbellIdHash, const char* messag
-            {
-                logger::error("video ssrc in barbell UMM does not exist %u", _loggableId.c_str(), ssrc);
-            }
-            else if (videoStream->endpointIdHash.isSet() && videoStream->endpointIdHash.get() != item.endpointIdHash)
+            _activeMediaList->removeVideoParticipant(entry.first);
+            _engineStreamDirector->removeParticipant(entry.first);
+
+            for (const auto ssrc : item.oldSsrcs)
             {
-                // must remove as this ssrc has been remapped
-                _activeMediaList->removeVideoParticipant(videoStream->endpointIdHash.get());
-                _engineStreamDirector->removeParticipant(videoStream->endpointIdHash.get());
-                videoStream->endpointIdHash.clear();
-                videoStream->endpointId.clear();
+                auto* videoStream = barbell->videoSsrcMap.getItem(ssrc);
+                if (videoStream)
+                {
+                    videoStream->endpointIdHash.clear();
+                    videoStream->endpointId.clear();
+                }
             }
         }
     }
 
-    for (auto& item : audioSsrcs)
+    // add videos
+    for (const auto& entry : videoSsrcs)
     {
-        for (size_t i = 0; i < item.ssrcs.size(); ++i)
+        const auto& item = entry.second;
+        if (item.hasChanged() && !item.newSsrcs.empty())
         {
-            const auto ssrc = item.ssrcs[i];
-            auto* audioStream = barbell->audioSsrcMap.getItem(ssrc);
-            if (!audioStream)
+            SimulcastStream primary;
+            utils::Optional<SimulcastStream> secondary;
+            uint32_t streamIndex = 0;
+            for (const auto ssrc : item.newSsrcs)
             {
-                logger::error("audio ssrc in barbell UMM does not exist %u", _loggableId.c_str(), ssrc);
+                auto* videoStream = barbell->videoSsrcMap.getItem(ssrc);
+                if (!videoStream)
+                {
+                    assert(false);
+                    logger::error("unannounced video ssrc %u", _loggableId.c_str(), ssrc);
+                    continue;
+                }
+                if (streamIndex++ == 0)
+                {
+                    primary = videoStream->stream;
+                }
+                else
+                {
+                    secondary.set(videoStream->stream);
+                }
+                videoStream->endpointIdHash.set(entry.first);
+                videoStream->endpointId.set(item.endpointId);
             }
-            else if (audioStream->endpointIdHash.isSet() && audioStream->endpointIdHash.get() != item.endpointIdHash)
+
+            _activeMediaList->addBarbellVideoParticipant(entry.first, primary, secondary, item.endpointId);
+            _engineStreamDirector->addParticipant(entry.first, primary, secondary.isSet() ? &secondary.get() : nullptr);
+
+            for (const auto& simLevel : utils::Span<SimulcastLevel>(primary.levels, primary.numLevels))
```

Comment:
I think this will became a regular pattern. Maybe it would be better to create a method on `SimulcastStream`  to return `utils::Span<SimulcastLevel>` (example,  `primary.getActiveLevels()`)